Allow for pullInterval of metrics to be configurable via ENV VAR PULL…

dnatic09 opened the pull request:
…_INTERVAL

<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Ensure you have added the unit tests for your changes.
2. Ensure you have included output of manual testing done in the Testing section.
3. Ensure number of lines of code for new or existing methods are within the reasonable limit.
4. Ensure your change works on existing clusters after upgrade.
5. If your mounting any new file or directory, make sure its not opening up any security attack vector for aws-vpc-cni-k8s modules.
6. If AWS apis are invoked, document the call rate in the description section.
7. If EC2 Metadata apis are invoked, ensure to handle stale information returned from metadata.
-->
**What type of PR is this?**
feature

<!--
Add one of the following:
bug
cleanup
documentation
feature
-->

**Which issue does this PR fix**:
[Allows for pullInterval to be configurable](https://github.com/aws/amazon-vpc-cni-k8s/issues/2171)

**What does this PR do / Why do we need it**:
Allows for pullInterval to be configurable via an environment variable

**If an issue # is not available please add repro steps and logs from IPAMD/CNI showing the issue**:


**Testing done on this change**:
<!--
output of manual testing/integration tests results and also attach logs
showing the fix being resolved
-->
Notice that the modified log entry shows what the pullInterval is to demonstrate the ENV VAR is parsing correctly.

***Manual: No ENV Vars***
```
GOROOT=/opt/homebrew/Cellar/go/1.19.4/libexec #gosetup
GOPATH=/Users/dnatic/go #gosetup
/opt/homebrew/Cellar/go/1.19.4/libexec/bin/go build -o /private/var/folders/7h/y0_07xp15tb2fxy0bsr7v9qm0000gn/T/GoLand/___go_build_main_go /Users/dnatic/workspace/aws/amazon-vpc-cni-k8s/cmd/cni-metrics-helper/main.go #gosetup
/private/var/folders/7h/y0_07xp15tb2fxy0bsr7v9qm0000gn/T/GoLand/___go_build_main_go
2022-12-20 14:18:19.256157 -0500 EST m=+0.012941168 write error: can't make directories for new logfile: mkdir /host: read-only file system
2022-12-20 14:18:19.256319 -0500 EST m=+0.013102835 write error: can't make directories for new logfile: mkdir /host: read-only file system
{"level":"info","ts":"2022-12-20T14:18:19.260-0500","caller":"cni-metrics-helper/main.go:46","msg":"Constructed new logger instance"}
{"level":"info","ts":"2022-12-20T14:18:19.260-0500","caller":"runtime/proc.go:250","msg":"Starting CNIMetricsHelper. Sending metrics to CloudWatch: true, LogLevel Debug, pullInterval 30"}
{"level":"error","ts":"2022-12-20T14:18:50.069-0500","caller":"runtime/proc.go:250","msg":"Failed to initialize kube client mapper: Get \"https://CC5534A988E436850F100114397BF816.gr7.eu-west-1.eks.amazonaws.com/api?timeout=32s\": dial tcp 10.188.68.82:443: i/o timeout"}

Process finished with the exit code 1

```

***Manual: PULL_INTERVAL=60***
```
Cellar/go/1.19.4/libexec #gosetup
GOPATH=/Users/dnatic/go #gosetup
/opt/homebrew/Cellar/go/1.19.4/libexec/bin/go build -o /private/var/folders/7h/y0_07xp15tb2fxy0bsr7v9qm0000gn/T/GoLand/___1go_build_main_go /Users/dnatic/workspace/aws/amazon-vpc-cni-k8s/cmd/cni-metrics-helper/main.go #gosetup
/private/var/folders/7h/y0_07xp15tb2fxy0bsr7v9qm0000gn/T/GoLand/___1go_build_main_go
2022-12-20 14:20:18.623736 -0500 EST m=+0.014495085 write error: can't make directories for new logfile: mkdir /host: read-only file system
2022-12-20 14:20:18.624018 -0500 EST m=+0.014776126 write error: can't make directories for new logfile: mkdir /host: read-only file system
{"level":"info","ts":"2022-12-20T14:20:18.627-0500","caller":"cni-metrics-helper/main.go:46","msg":"Constructed new logger instance"}
{"level":"info","ts":"2022-12-20T14:20:18.627-0500","caller":"runtime/proc.go:250","msg":"Starting CNIMetricsHelper. Sending metrics to CloudWatch: true, LogLevel Debug, pullInterval 60"}
{"level":"error","ts":"2022-12-20T14:20:49.549-0500","caller":"runtime/proc.go:250","msg":"Failed to initialize kube client mapper: Get \"https://CC5534A988E436850F100114397BF816.gr7.eu-west-1.eks.amazonaws.com/api?timeout=32s\": dial tcp 10.188.72.64:443: i/o timeout"}

Process finished with the exit code 1
```

***Manual: PULL_INTERVAL=bad-value-fail***
```
GOROOT=/opt/homebrew/Cellar/go/1.19.4/libexec #gosetup
GOPATH=/Users/dnatic/go #gosetup
/opt/homebrew/Cellar/go/1.19.4/libexec/bin/go build -o /private/var/folders/7h/y0_07xp15tb2fxy0bsr7v9qm0000gn/T/GoLand/___2go_build_main_go /Users/dnatic/workspace/aws/amazon-vpc-cni-k8s/cmd/cni-metrics-helper/main.go #gosetup
/private/var/folders/7h/y0_07xp15tb2fxy0bsr7v9qm0000gn/T/GoLand/___2go_build_main_go
2022-12-20 14:21:34.703325 -0500 EST m=+0.014378918 write error: can't make directories for new logfile: mkdir /host: read-only file system
2022-12-20 14:21:34.703531 -0500 EST m=+0.014585293 write error: can't make directories for new logfile: mkdir /host: read-only file system
{"level":"info","ts":"2022-12-20T14:21:34.707-0500","caller":"cni-metrics-helper/main.go:46","msg":"Constructed new logger instance"}
{"level":"fatal","ts":"2022-12-20T14:21:34.707-0500","caller":"runtime/proc.go:250","msg":"pullInterval (bad-value-fail) format invalid. Integer required. Expecting seconds: strconv.Atoi: parsing \"bad-value-fail\": invalid syntax"}

Process finished with the exit code 1
```

**Automation added to e2e**:
<!-- 
Test case added to lib/integration.sh 
If no, create an issue with enhancement/testing label
-->
There are no tests in the `main` package. Not applicable

**Will this PR introduce any new dependencies?**:
<!-- 
e.g. new EC2/K8s API, IMDS API, dependency on specific kernel module/version or binary in container OS.
-->
No

**Will this break upgrades or downgrades. Has updating a running cluster been tested?**:
No

**Does this change require updates to the CNI daemonset config files to work?**:
<!--
If this change does not work with a "kubectl patch" of the image tag, please explain why.
-->
No

**Does this PR introduce any user-facing change?**:
<!--
If yes, a release note update is required:
Enter your extended release note in the block below. If the PR requires additional actions
from users switching to the new release, include the string "action required".
-->
No

```release-note

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

dnatic09 commented:
Can any maintainer/author comment on the Docker images failing? They are failing to pull with a 400 status code. I am not sure I would be able to fix that.

jayanthvn commented:
It might be a temporary error. I have restarted the workflow. Thanks for notifying.

dnatic09 commented:
> It might be a temporary error. I have restarted the workflow. Thanks for notifying.

It unfortunately failed again.

jdn5126 commented on cmd/cni-metrics-helper/main.go:
```diff
@@ -80,6 +81,16 @@ func main() {
 		}
 	}
 
+	pullIntervalEnv, found := os.LookupEnv("PULL_INTERVAL")
```
nit: Can we use a more descriptive name, maybe `METRIC_PULL_INTERVAL` or `METRIC_PULL_INTERVAL_SECS`?

dnatic09 replied:
I am more than happy to switch the ENV VAR name. I used this name to preserve the golang var name as much as possible. I'll switch it now.

dnatic09 commented:
@jdn5126 , @haouc ;
My branch is now out-of-date with `master`. May I rebase against `master` and force-push my branch?

jdn5126 commented:
> @jdn5126 , @haouc ; My branch is now out-of-date with `master`. May I rebase against `master` and force-push my branch?

Yeah definitely, and this change looks good to me, so we will just need to wait until build passes. Just wanted to make sure you were aware that there are two timing metrics here:
1. `METRIC_PULL_INTERVAL` - rate at which agent gets and aggregates metrics for publishing (default 30s). The lower the value, the more CPU is consumed.
2. publish interval (https://github.com/aws/amazon-vpc-cni-k8s/blob/master/pkg/publisher/publisher.go#L34) - rate at which local metrics are pushed to CloudWatch (default 60s).

jdn5126 commented:
If `METRIC_PULL_INTERVAL` > `publish interval`, you will repeatedly see this log: https://github.com/aws/amazon-vpc-cni-k8s/blob/master/pkg/publisher/publisher.go#L178

dnatic09 commented:
> > @jdn5126 , @haouc ; My branch is now out-of-date with `master`. May I rebase against `master` and force-push my branch?
> 
> Yeah definitely, and this change looks good to me, so we will just need to wait until build passes. Just wanted to make sure you were aware that there are two timing metrics here:
> 
> 1. `METRIC_PULL_INTERVAL` - rate at which agent gets and aggregates metrics for publishing (default 30s). The lower the value, the more CPU is consumed.
> 2. publish interval (https://github.com/aws/amazon-vpc-cni-k8s/blob/master/pkg/publisher/publisher.go#L34) - rate at which local metrics are pushed to CloudWatch (default 60s).

Oh, no. I was trying to modify the publishInterval to CloudWatch. Let me update the branch. I will also rebase.

jdn5126 commented:
@dnatic09 both being configurable seems reasonable to me, as you really need both to achieve your goal. The `publish interval` should be 2x the `pull interval`, so if you want to use one environment variable and just multiply it by 2, I am fine with that. You can just pass the `publish interval` as an argument to `publisher.New` - https://github.com/aws/amazon-vpc-cni-k8s/blob/master/cmd/cni-metrics-helper/main.go#L119

jdn5126 commented:
> @dnatic09 both being configurable seems reasonable to me, as you really need both to achieve your goal. The `publish interval` should be 2x the `pull interval`, so if you want to use one environment variable and just multiply it by 2, I am fine with that. You can just pass the `publish interval` as an argument to `publisher.New` - https://github.com/aws/amazon-vpc-cni-k8s/blob/master/cmd/cni-metrics-helper/main.go#L119

Or actually `publisher.Start` - https://github.com/aws/amazon-vpc-cni-k8s/blob/master/cmd/cni-metrics-helper/main.go#L123

dnatic09 commented:
> > @dnatic09 both being configurable seems reasonable to me, as you really need both to achieve your goal. The `publish interval` should be 2x the `pull interval`, so if you want to use one environment variable and just multiply it by 2, I am fine with that. You can just pass the `publish interval` as an argument to `publisher.New` - https://github.com/aws/amazon-vpc-cni-k8s/blob/master/cmd/cni-metrics-helper/main.go#L119
> 
> Or actually `publisher.Start` - https://github.com/aws/amazon-vpc-cni-k8s/blob/master/cmd/cni-metrics-helper/main.go#L123

Done. I updated Publisher to always use double the pull interval. I also rebased against `master` again since there was another merge.

@jdn5126 , thank you for your continued support.

jdn5126 commented on pkg/publisher/publisher.go:
```diff
@@ -136,10 +133,11 @@ func New(ctx context.Context, region string, clusterID string, log logger.Logger
+// Start is used to set up the monitor loop
+func (p *cloudWatchPublisher) Start(publishInterval int) {
+	p.log.Infof("Starting monitor loop for CloudWatch publisher with interval (%d)", publishInterval)
+	var publishIntervalDuration = time.Second * time.Duration(publishInterval)
```
nit: no need to cast int to `time.Duration`. And can you also update the log to something like "with push interval of %d seconds"?

dnatic09 replied:
Completed in the latest commit

jdn5126 commented on cmd/cni-metrics-helper/main.go:
```diff
@@ -120,15 +131,15 @@ func main() {
 			log.Fatalf("Failed to create publisher: %v", err)
 		}
-		go cw.Start()
+		var publishInterval = pullInterval * 2
+		go cw.Start(publishInterval)
```
bit nit, but it is a bit more golang to do `publishInterval := pullInterval * 2` or to just pass `pullInterval * 2` to function

dnatic09 replied:
Completed in the latest commit

Sorry. My language of choice is Java. My Java-like expressiveness is showing through :)

jdn5126 commented on cmd/cni-metrics-helper/main.go:
```diff
@@ -80,6 +81,16 @@ func main() {
 		}
 	}
 
+	pullIntervalEnv, found := os.LookupEnv("METRIC_PULL_INTERVAL")
```
nit: sorry for going back on this, but I think the word "PULL" might be confusing to users. Can we do `METRIC_UPDATE_INTERVAL`? When we finally add a README for this helm chart, we can then make it clear that this is the rate at which metrics are updated, and they are published to CloudWatch at 2*update rate

dnatic09 replied:
Completed in the latest commit

jdn5126 commented:
> > > @dnatic09 both being configurable seems reasonable to me, as you really need both to achieve your goal. The `publish interval` should be 2x the `pull interval`, so if you want to use one environment variable and just multiply it by 2, I am fine with that. You can just pass the `publish interval` as an argument to `publisher.New` - https://github.com/aws/amazon-vpc-cni-k8s/blob/master/cmd/cni-metrics-helper/main.go#L119
> > 
> > 
> > Or actually `publisher.Start` - https://github.com/aws/amazon-vpc-cni-k8s/blob/master/cmd/cni-metrics-helper/main.go#L123
> 
> Done. I updated Publisher to always use double the pull interval. I also rebased against `master` again since there was another merge.
> 
> @jdn5126 , thank you for your continued support.

Of course, thank you for doing this! A few nits, sorry, and then I am happy to approve.

jdn5126 commented:
I will also kick off a GitHub action to run the metrics-helper integration test once this passes build

dnatic09 commented:
@jdn5126 , pushed with all of your suggestions.

jdn5126 commented on cmd/cni-metrics-helper/main.go:
```diff
@@ -88,7 +99,7 @@ func main() {
 	clusterID, _ := os.LookupEnv("AWS_CLUSTER_ID")
 
-	log.Infof("Starting CNIMetricsHelper. Sending metrics to CloudWatch: %v, LogLevel %s", options.submitCW, logConfig.LogLevel)
+	log.Infof("Starting CNIMetricsHelper. Sending metrics to CloudWatch: %v, LogLevel %s, pullInterval %d", options.submitCW, logConfig.LogLevel, pullInterval)
```
looks like this needs to be updated to new var name: `metricUpdateInterval`

dnatic09 replied:
My apologies. Updated.

jdn5126 commented:
Darn, looks like I was wrong. Unit test failed with: `Error: pkg/publisher/publisher.go:139:29: invalid operation: time.Second * publishInterval (mismatched types time.Duration and int)`. Can you run `make` and `make check-format` locally after fixing?

dnatic09 commented:
@jdn5126 , fixed. `make` runs fine. I get this when I run `make check-format`:

(M1 OS X)
```
% make check-format

xargs: goimports: No such file or directory
ERROR: goimports not installed
make: *** [format] Error 1
```

dnatic09 commented:
@jdn5126 , I can easily make this change to `Mock_Publisher.go`. Is this what it is asking for?

```
// Start mocks base method
func (m *MockPublisher) Start(publishInterval int) {
	m.ctrl.Call(m, "Start")
}
```

jdn5126 commented:
> @jdn5126 , I can easily make this change to `Mock_Publisher.go`. Is this what it is asking for?
> 
> ```
> // Start mocks base method
> func (m *MockPublisher) Start(publishInterval int) {
> 	m.ctrl.Call(m, "Start")
> }
> ```

Ah, sorry, how could I forget tests 🤦🏻 . Yeah, you will need to update the mock definition to:
```
// Start mocks base method
func (m *MockPublisher) Start(publishInterval int) {
	m.ctrl.Call(m, "Start", publishInterval)
}
```
It doesn't look to me like any test code needs to be updated. You can run `go test` from `pkg/publisher` directory and from `cmd/cni-metrics-helper/metrics` directory to confirm that the relevant tests pass.

For the `make check-format` error, you can install `goimports` with `go install golang.org/x/tools/cmd/goimports@latest`

dnatic09 commented:
I am not sure what is happening with `goimports`.
```
(base) dnatic@Daniels-MBP amazon-vpc-cni-k8s % pwd
/Users/dnatic/workspace/aws/amazon-vpc-cni-k8s
(base) dnatic@Daniels-MBP amazon-vpc-cni-k8s % go install golang.org/x/tools/cmd/goimports@latest
(base) dnatic@Daniels-MBP amazon-vpc-cni-k8s % make check-format                                 
xargs: goimports: No such file or directory
ERROR: goimports not installed
make: *** [format] Error 1
(base) dnatic@Daniels-MBP amazon-vpc-cni-k8s % 
```

I fixed the broken test. All tests passed with `go test` in the directories you specified.

jdn5126 commented:
> I am not sure what is happening with `goimports`.
> 
> ```
> (base) dnatic@Daniels-MBP amazon-vpc-cni-k8s % pwd
> /Users/dnatic/workspace/aws/amazon-vpc-cni-k8s
> (base) dnatic@Daniels-MBP amazon-vpc-cni-k8s % go install golang.org/x/tools/cmd/goimports@latest
> (base) dnatic@Daniels-MBP amazon-vpc-cni-k8s % make check-format                                 
> xargs: goimports: No such file or directory
> ERROR: goimports not installed
> make: *** [format] Error 1
> (base) dnatic@Daniels-MBP amazon-vpc-cni-k8s % 
> ```
> 
> I fixed the broken test. All tests passed with `go test` in the directories you specified.

It should install in your `$GOPATH/pkg/mod location`, so something like `~/go/pkg/mod/golang.org/x/tools...`, so it is possible that your $GOPATH is off, but not a big deal, up to you if you want to fix it.

The typical directory organization would be to checkout the repo in`$GOPATH/src/github.com/amazon-vpc-cni-k8s/...`, and then it would know to look for modules in `$GOPATH/pkg/mod/...`

dnatic09 commented:
Okay. It is definitely my fault then with the `GOPATH`. I do not have it in the proper directory structure.

dnatic09 commented:
Unit tests pass now :)

jdn5126 commented:
> Unit tests pass now :)
> 
> EDIT: All checks pass :)

Merged, thank you for the contribution, and for following through!

dnatic09 commented:
@jdn5126 , thank you again for supporting the enhancement. I look forward to additional contributions.
